### DISCUSSION TITLE
[PorousFlow] - Single Phase - Multicomponent flow - density/ viscosity

### DISCUSSION THREAD

**Discussion by 1runer** [idaholab/moose#23596] [Q&A General] [answered]
Dear all,

I am trying to simulate a single phase - two component system as mixed flow of a fluid and with a higher density fluid.
Since I assume that I have a component with a mass fraction > 0.1 , I consider that rho and mu changes due to the mass concentration and I also have to consider this in the flow equations. 

I have therefore written my own material inspired by the material PorousFlowBrine, which calculates rho/mu as a function of p_T_x by means of a FluidProperties UserObject. Correspondingly I added also a MultiComponentFluidProperties Userobject that contains the mixture properties equations for this and which is coupled to a SinglePhase object e.g. for water. 

So I can successfully derive the mixture properties for rho/mu which I successfully solve using the Kernels of PorousFlowMassTimeDerivative and PorousFlowAdvectiveFlux for both variables of mass_fraction0 and pore_pressure. 

Since of course the convergence through the density and viscosity contrasts is not too good, I assume that I have to include their derivatives wrt X as well. To transfer these from the user object into the material and to fill the local (*_ddensity_dx)[_qp] with it, worked so far. Now I'm stuck on how to get them into the corresponding flux kernels. I do not understand how the corresponding derivatives (drho_dX and dmu_dX) in the MaterialProperties vectors are structured and filled in dependence of the phase and components. The structure of PorousFlow is not simple in this respect. 

Therefore I wanted to ask if there is any further documentation or good examples (code as well as input files)? Or do I have a mistake in thinking and don't need these properties at all?

**Comment by cpgr** [comment, marked as answer]
Well, this is a little embarrassing. I was about to say that it should be easy but just had a quick look and I don't think we are actually passing those derivatives wrt mass fraction from porousflowbrine...

I'll fix this up this weekend so that it should just work for your case. Thanks for bringing this to our attention!

**Reply by cpgr** [reply]
Hopefully #23611 fixes this for you!

**Reply by cpgr** [reply]
#23611 is in so this should work for you now using a `PorousFlowMultiComponentFluid` material.

**Comment by 1runer** [comment]
Hi Chris,

thanks for the update and also finally to get it merged.
Yes, that¨s what I need.

Best

**Comment by josebastiase** [comment]
Hi @1runer,

Your UserObject may be really helpful for me. Any chance that you can merge it to MOOSE or share it? 

Thanks!
